packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -0,0 +1,223 @@
+  final TextSelectionHandleBuilder builder;
+  TextSelection selection;
+
+  /// A pair of handles. If this is non-null, there are always 2, though the
+  /// second is hidden when the selection is collapsed.
+  List<OverlayEntry> _handles;
+
+  /// Shows the handles by inserting them into the [context]'s overlay.
+  void show(BuildContext context, { Widget debugRequiredFor }) {
+    assert(_handles == null);
+    _handles = [
+      new OverlayEntry(builder: (BuildContext c) => _buildOverlay(c, 0)),
+      new OverlayEntry(builder: (BuildContext c) => _buildOverlay(c, 1)),
+    ];
+    Overlay.of(context, debugRequiredFor: debugRequiredFor).insertAll(_handles);
+  }
+
+  /// Updates the handles after the [selection] has changed.
+  void update(TextSelection newSelection) {
+    selection = newSelection;
+    _handles[0].markNeedsBuild();
+    _handles[1].markNeedsBuild();
+  }
+
+  /// Hides the handles.
+  void hide() {
+    _handles[0].remove();
+    _handles[1].remove();
+    _handles = null;
+  }
+
+  Widget _buildOverlay(BuildContext context, int which) {
+    if (selection.isCollapsed && which == 1)
+      return new OffStage(child: null);  // hide the second handle when collapsed
+    return new _TextSelectionHandleOverlay(
+      onSelectionHandleChanged: _handleSelectionHandleChanged,
+      renderObject: renderObject,
+      selection: selection,
+      builder: builder,
+      position: _positions[which]
+    );
+  }
+
+  void _handleSelectionHandleChanged(TextSelection newSelection) {
+    if (onSelectionHandleChanged != null)
+      onSelectionHandleChanged(newSelection);
+    update(selection);
+  }
+}
+
+/// This widget represents a single draggable text selection handle.
+class _TextSelectionHandleOverlay extends StatefulWidget {
+  _TextSelectionHandleOverlay({
+    Key key,
+    this.selection,
+    this.position,
+    this.renderObject,
+    this.onSelectionHandleChanged,
+    this.builder
+  }) : super(key: key);
+
+  final TextSelection selection;
+  final _TextSelectionHandlePosition position;
+  final RenderEditableLine renderObject;
+  final ValueChanged<TextSelection> onSelectionHandleChanged;
+  final TextSelectionHandleBuilder builder;
+
+  @override
+  _TextSelectionHandleOverlayState createState() => new _TextSelectionHandleOverlayState();
+}
+
+class _TextSelectionHandleOverlayState extends State<_TextSelectionHandleOverlay> {
+  Point _dragPosition;
+  void _handleDragStart(Point position) {
+    _dragPosition = position;
+  }
+
+  void _handleDragUpdate(double delta) {
+    _dragPosition += new Offset(delta, 0.0);
+    TextPosition position = config.renderObject.getPositionForPoint(_dragPosition);
+
+    if (config.selection.isCollapsed) {
+      config.onSelectionHandleChanged(new TextSelection.fromPosition(position));
+      return;
+    }
+
+    TextSelection newSelection;
+    switch (config.position) {
+      case _TextSelectionHandlePosition.start:
```

Comment:
I'd probably use base/extent for these enum values because they map down to baseOffset and extentOffset, respectively, in TextSelection, but what you've got is also fine.